From 1b3ac595d34d549c7d36b29cd8ee56c0cb733121 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 20 Aug 2025 17:01:20 +0800
Subject: [PATCH 233/236] mtk: hostapd: use ap_celanup_sta() to safely cleanup
 STA when receive AUTH from associated STA

The original function hostapd_ml_handle_disconnect() requires that
sta_info from the setup link exists; otherwise, it might cause
hostapd to crash.

The function ap_cleanup_sta() can safely handle such cases.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c | 2 +-
 src/ap/sta_info.c   | 2 +-
 src/ap/sta_info.h   | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 02c1d5d5b..bf1c4a34a 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3453,7 +3453,7 @@ static void handle_auth(struct hostapd_data *hapd,
 				   "receive AUTH from associated STA "MACSTR
 				   ", do deauthentication\n",
 				   MAC2STR(sa));
-			hostapd_ml_handle_disconnect(hapd, sta, mgmt, false);
+			ap_cleanup_sta(hapd, sa);
 			hostapd_drv_sta_remove(hapd, sa);
 			resp = WLAN_STATUS_UNSPECIFIED_FAILURE;
 			sta = NULL;
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 6bf11cec4..e32c491c9 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -2043,7 +2043,7 @@ bool ap_sta_in_list(struct hostapd_data *hapd, struct sta_info *sta) {
 	return false;
 }
 
-void ap_cleanup_sta(struct hostapd_data *hapd, u8 *addr) {
+void ap_cleanup_sta(struct hostapd_data *hapd, const u8 *addr) {
 	struct hostapd_data *h;
 	struct sta_info *sta;
 
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index 707f1e7ab..ff093ae56 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -437,6 +437,6 @@ void hostapd_free_link_stas(struct hostapd_data *hapd);
 void clear_wpa_sm_for_each_partner_link(struct hostapd_data *hapd,
 					struct sta_info *psta);
 bool ap_sta_in_list(struct hostapd_data *hapd, struct sta_info *sta);
-void ap_cleanup_sta(struct hostapd_data *hapd, u8 *addr);
+void ap_cleanup_sta(struct hostapd_data *hapd, const u8 *addr);
 
 #endif /* STA_INFO_H */
-- 
2.45.2

